### PR TITLE
add support for wasm compile targets

### PR DESCRIPTION
**Pull request by magicxyyz**
part of NIT-2631

This PR: 
* changes wasm store schema to support storing assembly for multiple targets
* adds support for multiple targets in StateDB, including recording UserWasm for specific targets

### CODE REVIEW

File: core/state/database_arbitrum.go
Comment (magicxyyz) [inline core/state/database_arbitrum.go:11]: this cache is currently broken - the cache can be populated with asms for different targets

I'll either cache whole asmMap for given moduleHash OR only use cache when targetName == local target name.

File: core/state/statedb_arbitrum.go
Comment (tsahee) [inline core/state/statedb_arbitrum.go:243]: add here a small condition that if len(targetNames) ==0 -> do nothing and return.
Eventually, that's what we'll use to avoid unnecessary work when recording.

File: core/rawdb/accessors_state_arbitrum.go
Comment (tsahee) [inline core/rawdb/accessors_state_arbitrum.go:28]: not 100% certain - but I think maybe TargetArm and TargetAsm should be identical to the runtime.GOARCH on arm64/x86 architectures, which will be compatible with some exiting code

**Review by tsahee — COMMENTED** [review]
very good.
tiny suggestions

File: core/rawdb/accessors_state_arbitrum.go
Comment (magicxyyz) [reply core/rawdb/accessors_state_arbitrum.go]: I used GOARCH strings + I renamed the constants to TargetArm64 and TargetAmd64 to make it more consistent

File: core/rawdb/accessors_state_arbitrum.go
Comment (gligneul) [inline core/rawdb/accessors_state_arbitrum.go:94]: Should we log the error if it is something other than not found?

File: core/state/statedb_arbitrum.go
Comment (gligneul) [inline core/state/statedb_arbitrum.go:114]:
```suggestion
			err = errors.Join(fmt.Errorf("failed to read activated asm from database for target %v and module %v: %w", target, moduleHash, dbErr), err)
```

**Review by gligneul — COMMENTED** [review]
Looks good, just minor comments.

File: core/vm/interface.go
Comment (ganeshvanahalli) [inline core/vm/interface.go:32]:
```suggestion
	ActivateWasm(moduleHash common.Hash, asmMap map[string][]byte)
```

File: core/rawdb/accessors_state_arbitrum.go
Comment (magicxyyz) [reply core/rawdb/accessors_state_arbitrum.go]: I changed ReadWasmSchemaVersion to return error and added handling it on nitro side - we already have a regex there to check if the error is "not found" error (unfortunately, pebble and leveldb return different errors, maybe we shouold patch that in future?)

**Review by tsahee — APPROVED** [review]
LGTM

File: core/state/statedb_arbitrum.go
Comment (PlasmaPower) [inline core/state/statedb_arbitrum.go:115]: Side note, but I had no idea `errors.Join` existed, that's really helpful!

**Review by PlasmaPower — CHANGES_REQUESTED** [review]

**Review by gligneul — APPROVED** [review]

**Review by PlasmaPower — COMMENTED** [review]
LGTM (pending nitro PR approval)

**Review by PlasmaPower — APPROVED** [review]

**Review by gligneul — APPROVED** [review]

**Pull request merged**
